### PR TITLE
Create a separate test library for each module/app

### PR DESCRIPTION
**Pull request by brianmoose**
Mainly wanted to see if this breaks apps and to start a discussion on whether this is the way we want to go with this.

This creates a new test library for each module/app in the `test` directory. It will compile anything under `test` into the library. There should be a "TestApp" that registers test objects that live in that directory. `main.C` then creates a "TestApp" instead of "App" so that it can run with both test and non-test objects depending on the '--allow-test-objects' command line argument.

I have included stubs for each of the modules. Stubs are not required, I just added them to make sure things are working and for easier migration of test objects in the future. If the `test` directory doesn't exist, an empty test library is still created. @permcody seemed to think this was not a problem.
Also, the "TestApp" inherits from "App" rather than `MooseApp`. This seemed a clean way to get everything registered properly but @permcody says this is not the way to go. I will look into alternate ways to cleanly get everything registered.

`stochastic_tools` provides a fuller example of what all this would look like as I moved some obvious test objects into the `test` directory along with their tests.

### CODE REVIEW

File: framework/app.mk [hunk 1]
```diff
@@ -5,6 +5,8 @@
 # list of application-wide excluded source files
 excluded_srcfiles :=
 
+print-%  : ; @echo $* = $($*)
```
Comment (permcody) [inline framework/app.mk]: What exactly does this line do?

File: framework/app.mk [hunk 2]
```diff
@@ -81,8 +100,13 @@ app_deps     := $(patsubst %.$(obj-suffix), %.$(obj-suffix).d, $(objects)) \
                 $(patsubst %.C, %.$(obj-suffix).d, $(main_src)) \
                 $(ADDITIONAL_APP_DEPS)
 
+app_test_deps     := $(patsubst %.$(obj-suffix), %.$(obj-suffix).d, $(test_objects)) \
+                $(patsubst %.c, %.$(obj-suffix).d, $(test_csrcfiles))
 depend_dirs := $(foreach i, $(DEPEND_MODULES), $(MOOSE_DIR)/modules/$(i)/include)
 depend_dirs += $(APPLICATION_DIR)/include
+ifneq ($(wildcard $(APPLICATION_DIR)/test/include/*),)
```
Comment (permcody) [inline framework/app.mk]: I don't think this would work for multilevel includes (why don't really have many, if any of these).

**Review by permcody — COMMENTED** [review]
You'll need to add the module headers to all the new source files below. I'll do a full review of this after testing.

File: framework/app.mk (hunk 1 above)
Comment (brianmoose) [inline framework/app.mk]: Just a target used for debugging, I will remove it.
It allows you to do something like `make print-depend_dirs` and it prints out the variable `depend_dirs`.

File: framework/app.mk (hunk 2 above)
Comment (brianmoose) [inline framework/app.mk]: This is just a test to see if there is anything in `test/include` and if there is then add `test/include` to the list of directories that we use with "-I". I guess it doesn't hurt to add the directory regardless.

**Comment by brianmoose** [comment]
peacock should also be updated to have the option to turn off/on test objects.

**Comment by jessecarterMOOSE** [comment]
I have a few objects in my App that are only used for testing (i.e. pulling data out of UserObjects) so I'd be interested in using this. Can someone provide a demo on this capability?

**Comment by permcody** [comment]
This is still in development at the moment. We'll announce this on the mailing list when it's done (soon).

File: modules/stochastic_tools/test/tests/samples/tests [hunk 3]
```diff
@@ -3,10 +3,12 @@
     type = RunApp
     input = threads.i
     min_threads = 2
+    cli_args = --allow-test-objects
   [../]
   [./mpi]
     type = RunApp
     input = mpi.i
     min_parallel = 2
+    cli_args = --allow-test-objects
```
Comment (permcody) [inline modules/stochastic_tools/test/tests/samples/tests]: Make sure you add a new test harness option

**Comment by brianmoose** [comment]
The two heavy tests that failed are known failures and not due to this PR.

**Review by permcody — APPROVED** [review]
Hold onto your butts...

**Comment by permcody** [comment]
Fails are all not related to this PR.

**Pull request merged**